go: default to 1.16, use binary bootstrap for all platforms

bobrik opened the pull request:
###### Motivation for this change

This mirrors what Rust does and allows us to drop go1.4 from the process.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS (x86_64 and aarch64 with some changes)
   - [x] other Linux distributions (aarch64)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

bobrik committed:
go: use binary bootstrap for all platforms

This mirrors what Rust does and allows us to drop go1.4 from the process.

bobrik commented:
cc @NixOS/golang

zowoq commented on pkgs/development/compilers/go/1.14.nix:
```diff
@@ -11,8 +11,16 @@ let
 
-  go_bootstrap = callPackage ./bootstrap.nix {
-    inherit Security;
+  go_bootstrap = callPackage ./binary.nix {
```
You changed this to `binary.nix`, the others are `bootstrap.nix`.

bobrik replied:
Good catch, amended.

zowoq commented on pkgs/development/compilers/go/1.14.nix:
```diff
@@ -11,8 +11,16 @@ let
-  go_bootstrap = callPackage ./bootstrap.nix {
-    inherit Security;
+  go_bootstrap = callPackage ./binary.nix {
+    version = "1.13.15"; # Using prior version, just like Rust.
```
The `rust` part of the comment can be removed.
```suggestion
    version = "1.13.15"; # Using prior version.
```

zowoq commented on pkgs/development/compilers/go/bootstrap.nix:
```diff
@@ -1,17 +1,35 @@
 in
-  go_bootstrap
+stdenv.mkDerivation rec {
+  name = "go-${version}-${platform}-binary";
```
```suggestion
  name = "go-${version}-${platform}-bootstrap";
```

zowoq commented:
We'll probably want to add a script like the rust `print-hashes.sh` as well.

r-burns commented:
I get the same error as ofborg when sandboxing is enabled:
```
building
no Makefile, doing nothing
./make.bash: line 165: /nix/store/13x2ia597vbb3r2s9i1izjy34bl8kl12-go-bootstrap/share/go/bin/go: No such file or directory
Building Go cmd/dist using /nix/store/13x2ia597vbb3r2s9i1izjy34bl8kl12-go-bootstrap/share/go. ()
./make.bash: line 181: /nix/store/13x2ia597vbb3r2s9i1izjy34bl8kl12-go-bootstrap/share/go/bin/go: No such file or directory
builder for '/nix/store/3sg6r8655knadr04jnckl49p5cigrfdq-go-1.15.8.drv' failed with exit code 127
```

bobrik commented:
Turns out I disabled sandboxing in my VM and forgot about it. Added `patchelf` similar to Rust, let's see if it helps ofborg as much as it does locally.

zowoq commented on pkgs/development/compilers/go/1.15.nix:
```diff
@@ -12,7 +12,15 @@ let
 
   go_bootstrap = callPackage ./bootstrap.nix {
-    inherit Security;
+    version = "1.14.15"; # Using prior version
```
Does it need to be the prior version and is it documented anywhere?

bobrik replied:
It's what Rust does for Rust reasons, but I don't think we have to do it for Go:

* https://github.com/NixOS/nixpkgs/commit/ba9c71b999b1

Happy to change to the same version for simplicity, let me know what you think.

bobrik replied:
Amended to use the same version for now.

zowoq replied:
I can't find anything that says we can't use the same version. It does make adding support for new arches easier.

zowoq commented on pkgs/development/compilers/go/1.4.nix:
```diff
@@ -1,93 +0,0 @@
-  '';
-
-  patches = [
-    ./remove-tools-1.4.patch
```
This file can be removed.

bobrik replied:
Removed

bobrik commented:
@r-burns ofborg is happy now.

zowoq commented on pkgs/development/compilers/go/1.15.nix:
```diff
@@ -11,8 +11,19 @@ let
 
   inherit (lib) optionals optionalString;
 
+  version = "1.15.8";
```
I don't know that we'd want to update the bootstrap for every new patch version, also all the new tarballs will be added to the hydra cache.

We could do `bootstrap_version = ${lib.versions.majorMinor version}` or similar?

r-burns replied:
Even that might be too often, go doesn't require recent features to build the compiler so I assume it would only need to be updated when they make available new bootstrap binaries for an additional architecture.

bobrik replied:
Makes sense, updated to use `lib.versions.majorMinor`.

zowoq replied:
Maybe keep the existing ones at 1.13 and start using 1.16 to bootstrap from then?

bobrik replied:
Take a look at the latest revision and let me know what you think.

zowoq commented on pkgs/development/compilers/go/print-hashes.sh:
```diff
@@ -0,0 +1,26 @@
+for PLATFORM in "${PLATFORMS[@]}"
+do
+    URL="$BASEURL/go$VERSION.$PLATFORM.tar.gz"
+    SHA256=$(curl -sSfL $URL | shasum -a 256 | cut -d ' ' -f 1)
```
We could use the API for this: https://golang.org/dl/?mode=json&include=all

bobrik replied:
TIL, updated.

kalbasit commented on pkgs/development/compilers/go/binary.nix:
```diff
@@ -0,0 +1,40 @@
+    '')}
+  '' ;
+
+  phases = [ "unpackPhase" "patchPhase" "installPhase" ];
```
Why do we need to specify the phases?

bobrik replied:
Copy pasted these from `srcOnly`, which is what I used as a starting point locally.

It works without, so I've dropped it 👍

kalbasit reviewed (commented):
👏

zowoq commented on pkgs/development/compilers/go/1.15.nix:
```diff
@@ -11,9 +11,9 @@ let
-  };
+  version = "1.15.8";
+
+  go_bootstrap = callPackage ./bootstrap.nix { };
```
How do you intend this to work so we can keep 1.13 bootstrap for 1.15 and we add a 1.16 bootstrap for 1.16?

bobrik replied:
We can pass `version` here in inputs and check it in `bootstrap.nix` to select the appropriate bootstrap version.

If go1.16 can bootstrap everything all the same, is there a reason not to use it?

zowoq replied:
> If go1.16 can bootstrap everything

Does using 1.16 to bootstrap 1.15 work on all platforms?

kalbasit replied:
Over on the installing from source [instructions](https://golang.org/doc/install/source#bootstrapFromBinaryRelease) they say that we can use any version as bootstrap so I think that we can use any version of Go and we don't have to use one per version. IIRC the bootstrap is used to build the compiler then the compiler is then used to build itself anyway.

bobrik replied:
I don't know for sure (but suspect yes), that's why passing `version` is a backup plan.

zowoq replied:
Another reason to keep separate bootstraps for now: just updating the bootstrap to add `go_1_16` (without changing the default `go` version) will cause a rebuild of all go packages.

zowoq replied:
We could just wait to merge this after we've added `go_1_16` and made it the default?

bobrik replied:
> Another reason to keep separate bootstraps for now: just updating the bootstrap to add `go_1_16` (without changing the default go version) will cause a rebuild of all go packages.

I'm not sure why we need to update bootstrap to add `go_1_16` until go1.16 is the default.

The timeline I have in mind:

1. Merge this, rebuilding all Go packages.
2. Merge #107937, building packages for Go 1.16 (bootstrapped from Go 1.13).
3. Merge a PR that changes both bootstrap and default Go versions to 1.16, rebuilding all packages.

I don't really see any convenient scenarios with fewer than two full rebuilds between current situation and Go 1.16 being the default with this PR merged.

zowoq replied:
If we do 1 and 3 at the same time it's one rebuild.

bobrik replied:
I'm ok with it if you think that saving one rebuild is worth it.

kalbasit replied:
Why do we need to move the bootstrap to go 1.16 at all? 🤔

zowoq replied:
To bootstrap on `aarch64-darwin`.

bobrik replied:
We'll definitely need it for `aarch64-darwin`, since no release other than 1.16 can bootstrap it. Having the same version for everything else is probably nice for consistency.

Either way, that doesn't really change the number of rebuilds required with this PR.

kalbasit replied:
Ahh I see, yea that makes sense. I'm ok waiting till we get 1.16 merged before merging this. No harm keeping it aside for now...

bobrik replied:
@zowoq should I update this PR to 1.16 then?

kalbasit replied:
@bobrik since 1.16 was released then yes, please. I'll merge this as soon as #107937 is updated to 1.16 and merged so hydra rebuilds both together.

zowoq replied:
> I'll merge this as soon as #107937 is updated to 1.16

We still need to change the defaults to 1.16 here and test them first.

bobrik replied:
Updated (including `darwin-arm64`).

zowoq replied:
Please rebase on master and apply the bootstrap changes to `1.16.nix`.

To change the defaults to `go_1_16` you can cherry-pick https://github.com/zowoq/nixpkgs/commit/76e01e2b6d1744661cd0ed28525b5d9bfbac18fe or I can push it to this branch for you.

zowoq replied:
Sorry, wrong commit. Should have been https://github.com/zowoq/nixpkgs/commit/8f3a96cfd043704a6d227b13e867c9528b1e522e.

bobrik replied:
Rebased and cherry-picked.

zowoq committed:
go, buildGoModule, buildGoPackage: default to go 1.16

zowoq commented on pkgs/development/compilers/go/bootstrap.nix:
```diff
@@ -1,17 +1,15 @@
-    inherit Security;
+{ callPackage }:
+callPackage ./binary.nix {
+  version = "1.13";
```
```suggestion
  version = "1.16";
```

bobrik replied:
Oops, fixed.

Mic92 commented on pkgs/development/compilers/go/print-hashes.sh:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
```suggestion
#!/usr/bin/env nix-shell
#!nix-shell -p bash -p jq -p curl -i bash
```

bobrik replied:
I can open a separate PR to update both Go and Rust versions to do this.

zowoq commented:
Hmmm, seems like all of the go versions are failing to build on `x86_64-darwin`.

bobrik commented:
@zowoq is that what you're seeing?

```
$ GOROOT="$GOROOT_BOOTSTRAP" GOOS="" GOARCH="" GO111MODULE=off /nix/store/fbsbnj3ciy8mm3xrp7dxrqfhmslywaaw-go-bootstrap/share/go/bin/go version
Killed: 9

$ codesign -v /nix/store/fbsbnj3ciy8mm3xrp7dxrqfhmslywaaw-go-bootstrap/share/go/bin/go
/nix/store/fbsbnj3ciy8mm3xrp7dxrqfhmslywaaw-go-bootstrap/share/go/bin/go: invalid signature (code or signature have been modified)
In architecture: x86_64
```

The signature inside of the archive is valid.

bobrik committed:
go: do not strip bootstrap on darwin

bobrik commented:
Pushed 5799dc9804a to preserve the bootstrap signature on Darwin, which seems to fix the build for me locally.

bobrik commented:
@zowoq, @kalbasit, anything else I should do before we merge this?

zowoq commented:
I'm running `nixpkgs-review` and checking `nixosTests` as we need to go through the packages and (fix|pin) anything that fails like we did in https://github.com/NixOS/nixpkgs/pull/95885.

Also `staging-next` is very close to merging, we'll want to wait until after that happens to limit rebuilds.

zowoq committed:
dockerTools.tarsum: pin to go 1.15

zowoq committed:
v2ray: pin to go 1.15

zowoq committed:
prometheus: pin to go 1.15

zowoq commented:
@ofborg test docker-tools

zowoq commented:
> running `nixpkgs-review`

I think I've killed my builder. :(

@zimbatm are you able to run `nixpkgs-review` on this PR?

bobrik committed:
galene: pin to go 1.15

bobrik committed:
ibus-engines.bamboo: pin to go 1.15

bobrik committed:
nncp: pin to go 1.15

bobrik committed:
trillian: pin to go 1.15

bobrik committed:
grafana-loki: pin to go 1.15

bobrik committed:
step-ca: pin to go 1.15

bobrik committed:
coyim: pin to go 1.15

bobrik committed:
oauth2_proxy: pin to go 1.15

bobrik commented:
I checked new `aarch64-linux` failures from my local `nixpkgs-review` and pushed the fixes.

bobrik commented:
Firing up `x86_64-darwin` under `nixpkgs-review` to see if there's something new there.

zowoq commented:
Looks like the bootstrap changes breaks cross builds, e.g. `pkgsCross.aarch64-multiplatform` from the `docker-tools` test.

bobrik commented:
@zowoq does the following do the trick for you?

```diff
diff --git a/pkgs/development/compilers/go/binary.nix b/pkgs/development/compilers/go/binary.nix
index 9a0dc343546..169ba49c1f9 100644
--- a/pkgs/development/compilers/go/binary.nix
+++ b/pkgs/development/compilers/go/binary.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, version, hashes }:
+{ lib, stdenv, pkgsBuildHost, fetchurl, version, hashes }:
 let
   toGoKernel = platform:
     if platform.isDarwin then "darwin"
@@ -15,7 +15,7 @@ let

   toGoPlatform = platform: "${toGoKernel platform}-${toGoCPU platform}";

-  platform = toGoPlatform stdenv.hostPlatform;
+  platform = toGoPlatform stdenv.buildPlatform;
 in
 stdenv.mkDerivation rec {
   name = "go-${version}-${platform}-bootstrap";
@@ -26,7 +26,7 @@ stdenv.mkDerivation rec {
   };

   # We must preserve the signature on Darwin
-  dontStrip = stdenv.hostPlatform.isDarwin;
+  dontStrip = stdenv.isDarwin;

   installPhase = ''
     mkdir -p $out/share/go $out/bin
@@ -34,7 +34,7 @@ stdenv.mkDerivation rec {
     cp -r . $out/share/go
     ${lib.optionalString stdenv.isLinux (''
     patchelf \
-      --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
+      --set-interpreter $(cat ${pkgsBuildHost.stdenv.cc}/nix-support/dynamic-linker) \
       $out/bin/go
     '')}
   '' ;
```

I'm not a cross-compilation expert, but with this change builds seem to progress further for `nix-build docker-tools.nix`.

Let me know if you have a more concrete solution in mind.

r-burns commented:
I think doing this for each go version fixes cross:

```diff
diff --git a/pkgs/development/compilers/go/1.16.nix b/pkgs/development/compilers/go/1.16.nix
index ebba1a513b0..8267e9745dc 100644
--- a/pkgs/development/compilers/go/1.16.nix
+++ b/pkgs/development/compilers/go/1.16.nix
@@ -13,7 +13,7 @@ let
 
   version = "1.16";
 
-  go_bootstrap = callPackage ./bootstrap.nix { };
+  go_bootstrap = buildPackages.callPackage ./bootstrap.nix { };
 
   goBootstrap = runCommand "go-bootstrap" {} ''
     mkdir $out
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index da556c68b7c..5496ce4cf9a 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -10161,28 +10161,28 @@ in
     inherit (darwin.apple_sdk.frameworks) Security Foundation;
   } // lib.optionalAttrs (stdenv.cc.isGNU && stdenv.isAarch64) {
     stdenv = gcc8Stdenv;
-    buildPackages = buildPackages // { stdenv = gcc8Stdenv; };
+    buildPackages = buildPackages // { stdenv = buildPackages.gcc8Stdenv; };
   });
 ```

Because
* go-bootstrap is a native build input of go, so it needs to have an offset of -1. Otherwise, e.g. when building a go cross-compiler, it will try to make go-bootstrap a cross-compiler too
* have to specify buildPackages for the stdenv override, otherwise buildPackages.stdenv will be the same as pkgs.gcc8Stdenv

r-burns committed:
go: fix cross-compilation

Because:

* `go-bootstrap` is a native build input of go, so it needs to have
  an offset of -1. Otherwise, e.g. when building a go cross-compiler,
  it will try to make go-bootstrap a cross-compiler too.
* have to specify `buildPackages` for the `stdenv` override, otherwise
  `buildPackages.stdenv` will be the same as `pkgs.gcc8Stdenv`.

bobrik commented:
@r-burns thanks, pushed it to the PR! Hope you don't mind me attaching you as the author.

@zowoq, can you double check if it works for you? It breaks for me the same way `master` branch does: complaining about lack of KVM support in my VM.

zowoq commented:
@ofborg test docker-tools

r-burns commented:
Ah, I think I see the issue. Testing docker-tools.nix now and it seems to be working.

```diff
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 5496ce4cf9a..a8ad091cdff 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -282,7 +282,7 @@ in
   grsync = callPackage ../applications/misc/grsync { };
 
   dockerTools = callPackage ../build-support/docker {
-    go = go_1_15;
+    go = buildPackages.go_1_15;
     writePython3 = buildPackages.writers.writePython3;
   };
 ```

Args to callPackage aren't spliced so you have to manually specify the package set. This is like the main cause of cross-compilation papercuts in my experience, it's super annoying

bobrik commented:
@r-burns, I amended the latest commit 👍

bobrik commented:
@ofborg test docker-tools

bobrik commented:
Looks like it's a success for cross-compilation! 🎉

My old laptop is still churning though `x86_64-darwin` rebuilds: 300/1000 packages, one seems broken so far.

zimbatm commented:
nixpkgs-fmt is still running. So far I have seen the following issues:

```
builder for '/nix/store/55p1rlsmp3gh4xg2hdwkycf026qhfd8y-galene-0.2.drv' failed with exit code 1; last 10 log lines:
  === RUN   TestDuration
  --- PASS: TestDuration (0.00s)
  === RUN   TestTime
      rtptime_test.go:46: Expected 4000, got 121696
  --- FAIL: TestTime (0.01s)
  === RUN   TestNTP
  --- PASS: TestNTP (0.00s)
  FAIL
  FAIL      github.com/jech/galene/rtptime  0.034s
  FAIL
```

```
builder for '/nix/store/c7wl3r5b6rv1riqq94pf1v6s8ws1in13-ipfs-cluster-0.13.1.drv' failed with exit code 1; last 10 log lines:
  === RUN   TestClustersPeerRejoin
  2021-02-18T08:30:25.157Z     ERROR   cluster leaving cluster: crdt consensus component cannot remove peers
  --- PASS: TestClustersPeerRejoin (11.04s)
  === RUN   TestClusterSecretFormat
  --- PASS: TestClusterSecretFormat (0.00s)
  === RUN   TestSimplePNet
  --- PASS: TestSimplePNet (3.05s)
  FAIL
  FAIL   github.com/ipfs/ipfs-cluster    563.179s
  FAIL
```

```
builder for '/nix/store/rkbdbggpnb8rh4wdmx5amrhs0na2qjsd-python3.7-dask-gateway-server-0.9.0.drv' failed with exit code 1; last 10 log lines:
  setting SOURCE_DATE_EPOCH to timestamp 1604523679 of file dask-gateway-server-0.9.0/setup.cfg
  patching sources
  configuring
  no configure script, doing nothing
  building
  Executing setuptoolsBuildPhase
  running bdist_wheel
  running build
  running build_proxy
  go: cannot find main module; see 'go help modules'
```
Go 1.16 switches to go modules by default. Setting GO111MODULES=off should fix that last issue.

bobrik committed:
pythonPackages.dask-gateway-server: use GO111MODULE=off to build on go1.16

bobrik commented:
@zimbatm is that from `x86-64_linux`?

* `galene` should've been fixed in https://github.com/NixOS/nixpkgs/pull/113230/commits/67643e9fc5a5b701a1477457c03981b400123cac
* `ipfs-cluster` may need another try, since it worked for me on both `aarch64-linux` and `x86_64-darwin`
* `dask-gateway-server` is should've been fixed by https://github.com/NixOS/nixpkgs/pull/113230/commits/df7e2c46da6a717476c97d33e13826d351bc9c24 (just pushed)
* `prometheus` should've been fixed in https://github.com/NixOS/nixpkgs/pull/113230/commits/6ea9ec2672075d8b8eccd18b541f8024421d450e

I haven't seen any new failures on `x86_64-darwin` other than `dask-gateway-server` which is fixed now.

As soon as @zimbatm can confirm that the issues are fixed for them on the up-to-date branch, I think it's ready to go.

Staging has been merged into master as well: #112095.

zowoq commented:
`ipfs-cluster` fails on `x86-64_linux` but it also fails with `1.15`, IIRC it has always been flaky.

I think I've checked all of the `nixosTests` affected by this, haven't found any unresolved new failures.

bobrik commented:
Is it good to merge then or is there anything else to do?

zowoq commented:
I'll wait a bit to see if @zimbatm has any more failed builds but I think we're good to merge.

zowoq commented:
Thanks @bobrik and @r-burns!

mweinelt commented:
Why did this go into master and not staging given the amount of rebuilds this causes?

zowoq commented:
We send go updates to master, the number of rebuilds isn't that many and they are quick.

kvtb commented:
> Pushed [5799dc9](https://github.com/NixOS/nixpkgs/commit/5799dc9804a2ce0cebbc44a68b8ad0df02b834c6) to preserve the bootstrap signature on Darwin, which seems to fix the build for me locally.

Am I the only one using Linux?

The `go` binary does not need `patchelf` (at least on `x86_64-linux`).
Moreover, `patchelf` just fails, breaking whole `go` infrastructure.
